api/src/main/java/jakarta/data/repository/KeysetPageable.java

```diff
@@ -191,138 +190,95 @@ public interface Cursor {
     }
 
     /**
-     * Built-in implementation of Cursor.
+     * Returns the keyset values which are the starting point for
+     * keyset pagination.
+     *
+     * @return the keyset values.
      */
-    static final class CursorImpl implements Cursor {
-        /**
-         * Keyset values.
-         */
-        private final Object[] keyset;
-
-        /**
-         * Constructs a keyset cursor with the specified values.
-         *
-         * @param keyset keyset values.
-         * @throws IllegalArgumentException if no keyset values are provided.
-         */
-        CursorImpl(Object... keyset) {
-            this.keyset = keyset;
-            if (keyset == null || keyset.length == 0)
-                throw new IllegalArgumentException("No keyset values were provided.");
-        }
-
-        @Override
-        public boolean equals(Object o) {
-            return this == o || o != null
-                    && o.getClass() == getClass()
-                    && Arrays.equals(keyset, ((CursorImpl) o).keyset);
-        }
-
-        public Object getKeysetElement(int index) {
-            return keyset[index];
-        }
-
-        @Override
-        public int hashCode() {
-            return Arrays.hashCode(keyset);
-        }
-
-        public int size() {
-            return keyset.length;
-        }
-
-        @Override
-        public String toString() {
-            return new StringBuilder(27).append("Cursor@").append(Integer.toHexString(hashCode()))
-                            .append(" with ").append(keyset.length).append(" keys")
-                            .toString();
-        }
-    }
-
-    private final Cursor cursor;
-    private final Mode mode;
-
-    KeysetPageable(Pageable copyFrom, Mode mode, Cursor cursor) {
-        super(copyFrom.page, copyFrom.size, copyFrom.sorts);
-
-        if (cursor == null || cursor.size() == 0)
-            throw new IllegalArgumentException("No keyset values were provided.");
-
-        this.cursor = cursor;
-        this.mode = mode;
-    }
+    public Cursor getCursor();
 
+    /**
+     * Returns the direction of keyset pagination.
+     *
+     * @return the direction of keyset pagination.
+     */
+    public Mode getMode();
```

Comment:
for future planning do we want to use `getMode()` or `mode()`? When upgrading to Java 17 these could be records

Reply:
> When upgrading to Java 17 these could be records

Do you mean that the implementations of the Pageable/KeysetPageable interfaces could be records?  Or that Pageable/KeysetPageable themselves could become records?  I'm assuming you mean the former because the latter would prevent the possibility of provider-specific extensions that #65 aims to allow for.

I think it would be a good idea for the implementations to be records.  To accommodate that, we will need to do much more than change `getMode()` into `mode()`.  We also gave `getSize()` and `getPage()` and `getSorts()` and `getCursor()` that would need to be switched, and then some of those would collide with fluent-style `.size()` and `.page()`.  I used the fluent pattern to make it easier to add more properties in future spec versions.  We could either keep that and use different (longer) names for the fluent API, for example,

```
p = Pageable.ofSize(50)
            .withPage(5)
            .withSorts(Sort.asc("lastName"), Sort.asc("firstName), Sort.asc("id"))
            .afterKeyset(lname, fname, id);
page = p.page();
size = p.size();
sorts = p.sorts();
mode = p.mode();
cursor = p.cursor();
```

Or we could do away with the fluent pattern for the core properties and have something like,

```
p = Pageable.ofSize(50, 5, Sort.asc("lastName"), Sort.asc("firstName), Sort.asc("id"))
            .afterKeyset(lname, fname, id);
page = p.page();
size = p.size();
sorts = p.sorts();
mode = p.mode();
cursor = p.cursor();
```

The unfortunate part about the latter is the possibility of someone confusing the page size and the page number because both are just numbers in the parameter list, although I tried to make it more clear which is the size by including `Size` in the method name and placing the size parameter first.

Reply:
i mean the implementations could be records

Reply:
I was also puzzled why KeySetPageable should also be an interface and not a class implementing Pageable or (if Java 17+ was the minimum JDK) a record if you like?

Reply:
> I was also puzzled why KeySetPageable should also be an interface and not a class implementing Pageable or (if Java 17+ was the minimum JDK) a record if you like?

It looks like you opened #72 for a separate discussion of this question, so I'll let this be resolved there.

And I'll leave renaming the fluent methods to #67.